In [45]:
# Load packages
from costs_benefits_ssp.cb_calculate import CostBenefits
import pandas as pd
import os
import pathlib
import numpy as np


In [46]:
# Define paths
SSP_PATH = pathlib.Path(os.getcwd()).parent.parent.as_posix()
SSP_RUN  = os.path.join(SSP_PATH, "ssp_run_output/sisepuede_run_2025-10-29T19;49;25.722413")


CB_DEFAULT_DEFINITION_PATH = os.path.join(SSP_PATH,"cb/cb_cost_factors")
CB_DEFAULT_DEFINITION_FILE_PATH = os.path.join(CB_DEFAULT_DEFINITION_PATH, "cb_config_params.xlsx")

CB_OUTPUT = os.path.join(SSP_PATH,"cb/cb_results")

In [47]:
# Load data
ssp_data = pd.read_csv(os.path.join(SSP_RUN, "uganda.csv"))
att_primary = pd.read_csv(os.path.join(SSP_RUN,"ATTRIBUTE_PRIMARY.csv"))
att_strategy = pd.read_csv(os.path.join(SSP_RUN,"ATTRIBUTE_STRATEGY.csv"))
strategy_code_base = "BASE"

In [48]:
att_primary

,primary_id,design_id,strategy_id,future_id
0,0,0,0,0
1,70070,0,6004,0
2,74074,0,6008,0
3,75075,0,6009,0


In [49]:
# Instantiate CostBenefits object
cb = CostBenefits(ssp_data, att_primary, att_strategy, strategy_code_base)

# The export_db_to_excel method saves the initial configuration of the cost tables to an excel file.
# Each sheet represents a table in the cost and benefit program database.
# If the Excel file name is not given, the file will be saved with the default name cb_config_params.xlsx on the current python session.

#cb.export_db_to_excel(CB_DEFAULT_DEFINITION_FILE_PATH)

# Once that the excel file has been updated, we can reload it in order to update the cost factors database
cb.load_cb_parameters(CB_DEFAULT_DEFINITION_FILE_PATH)

The TX TX:FRST:INCREASE_SEQUESTRATION_NZ is missing on AttTransformationCode
The TX TX:LNDU:DEC_WETLAND_LOSS_NZ is missing on AttTransformationCode
The TX TX:LNDU:SET_WETLANDS_MINIMUM_NZ is missing on AttTransformationCode
The TX TX:SCOE:INC_EFFICIENCY_HEAT_NZ is missing on AttTransformationCode
The TX TX:FRST:INCREASE_SEQUESTRATION_NDC_2 is missing on AttTransformationCode
The TX TX:LNDU:DEC_WETLAND_LOSS_NDC_2 is missing on AttTransformationCode
The TX TX:LNDU:SET_WETLANDS_MINIMUM_NDC_2 is missing on AttTransformationCode
The TX TX:SCOE:INC_EFFICIENCY_HEAT_NDC_2 is missing on AttTransformationCode
The TX TX:FRST:INCREASE_SEQUESTRATION_NDC_25 is missing on AttTransformationCode
The TX TX:LNDU:DEC_WETLAND_LOSS_NDC_25 is missing on AttTransformationCode
The TX TX:LNDU:SET_WETLANDS_MINIMUM_NDC_25 is missing on AttTransformationCode
The TX TX:SCOE:INC_EFFICIENCY_HEAT_NDC_25 is missing on AttTransformationCode
Cargamos configuración de archivo excel
Se actualizó la base de datos


In [50]:
# Compute System Costs
results_system = cb.compute_system_cost_for_all_strategies()


************************************
*Strategy : PFLO:NZ (0/3)
************************************

---------Costs for: cb:wali:technical_cost:sanitation:unimp_rural.
La variable se evalúa en System Cost
---------Costs for: cb:wali:technical_cost:sanitation:imp_rural.
La variable se evalúa en System Cost
---------Costs for: cb:wali:technical_cost:sanitation:safeman_rural.
La variable se evalúa en System Cost
---------Costs for: cb:wali:technical_cost:sanitation:unimp_urban.
La variable se evalúa en System Cost
---------Costs for: cb:wali:technical_cost:sanitation:imp_urban.
La variable se evalúa en System Cost
---------Costs for: cb:wali:technical_cost:sanitation:safeman_urban.
La variable se evalúa en System Cost
---------Costs for: cb:wali:technical_cost:sanitation:omit_rural.
La variable se evalúa en System Cost
---------Costs for: cb:entc:technical_cost:electricity:capex.
La variable se evalúa en System Cost
---------Costs for: cb:entc:technical_cost:electricity:opex.
La variable

In [51]:
# Compute Technical Costs
results_tx = cb.compute_technical_cost_for_all_strategies()


************************************
*Strategy : PFLO:NZ (0/3)
************************************

---------Costs for: cb:inen:technical_cost:fuel_switch:lo_heat.
La variable se evalúa en Transformation Cost
---------Costs for: cb:inen:technical_savings:fuel_switch:lo_heat.
La variable se evalúa en Transformation Cost
---------Costs for: cb:inen:technical_cost:fuel_switch:hi_heat.
La variable se evalúa en Transformation Cost
---------Costs for: cb:inen:technical_savings:fuel_switch:hi_heat.
La variable se evalúa en Transformation Cost
---------Costs for: cb:inen:technical_cost:fuel_switch:all.
La variable se evalúa en Transformation Cost
---------Costs for: cb:inen:technical_savings:fuel_switch:all.
La variable se evalúa en Transformation Cost
---------Costs for: cb:inen:technical_cost:efficiency:X.
La variable se evalúa en Transformation Cost
---------Costs for: cb:trns:technical_cost:electrification:LDV.
La variable se evalúa en Transformation Cost
---------Costs for: cb:trns:tech

/Users/fabianfuentes/anaconda3/envs/ssp_uganda_env/lib/python3.11/site-packages/costs_benefits_ssp/cb_calculate.py:856: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  tmp["difference_variable"] = cb_orm.diff_var
/Users/fabianfuentes/anaconda3/envs/ssp_uganda_env/lib/python3.11/site-packages/costs_benefits_ssp/cb_calculate.py:857: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  tmp["difference_value"] = data_merged["difference"]
/Users/fabianfuentes/anaconda3/envs/ssp_uganda_env/lib/python3.11/site-packages/cos

---------Costs for: cb:ippu:technical_cost:abating_gases:other_fgases.
La variable se evalúa en Transformation Cost
---------Costs for: cb:fgtv:technical_cost:flaring:X.
La variable se evalúa en Transformation Cost
---------Costs for: cb:fgtv:technical_cost:leaks:X.
La variable se evalúa en Transformation Cost
---------Costs for: cb:waso:technical_cost:consumer_food_waste:X.
La variable se evalúa en Transformation Cost
---------Costs for: cb:waso:consumer_savings:consumer_food_waste:X.
La variable se evalúa en Transformation Cost
---------Costs for: cb:ccsq:technical_cost:direct_air_capture:X.
La variable se evalúa en Transformation Cost
---------Costs for: cb:lvst:technical_cost:ent_ferm_mgmt:X.
La variable se evalúa en Transformation Cost
---------Costs for: cb:agrc:technical_cost:rice_mgmt:X.
La variable se evalúa en Transformation Cost
---------Costs for: cb:agrc:technical_cost:producer_food_waste:X.
La variable se evalúa en Transformation Cost
---------Costs for: cb:agrc:technical

/Users/fabianfuentes/anaconda3/envs/ssp_uganda_env/lib/python3.11/site-packages/costs_benefits_ssp/cb_calculate.py:1067: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_output['variable_value_baseline'] = 0
/Users/fabianfuentes/anaconda3/envs/ssp_uganda_env/lib/python3.11/site-packages/costs_benefits_ssp/cb_calculate.py:1070: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_output['variable_value_pathway'] = data_merged["difference_value"]


---------Costs for: cb:inen:technical_savings:fuel_switch:all.
La variable se evalúa en Transformation Cost
---------Costs for: cb:inen:technical_cost:efficiency:X.
La variable se evalúa en Transformation Cost
---------Costs for: cb:trns:technical_cost:electrification:LDV.
La variable se evalúa en Transformation Cost
---------Costs for: cb:trns:technical_savings:electrification:LDV.
La variable se evalúa en Transformation Cost
---------Costs for: cb:trns:technical_cost:electrification:hdv_vehicles.
La variable se evalúa en Transformation Cost
---------Costs for: cb:trns:technical_savings:electrification:hdv_vehicles.
La variable se evalúa en Transformation Cost
---------Costs for: cb:trns:technical_cost:electrification:hdv_charger.
La variable se evalúa en Transformation Cost
---------Costs for: cb:trns:technical_cost:efficiency:non_electric.
La variable se evalúa en Transformation Cost
---------Costs for: cb:trns:technical_cost:efficiency:all.
La variable se evalúa en Transformation C

/Users/fabianfuentes/anaconda3/envs/ssp_uganda_env/lib/python3.11/site-packages/costs_benefits_ssp/cb_calculate.py:1067: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_output['variable_value_baseline'] = 0
/Users/fabianfuentes/anaconda3/envs/ssp_uganda_env/lib/python3.11/site-packages/costs_benefits_ssp/cb_calculate.py:1070: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_output['variable_value_pathway'] = data_merged["difference_value"]


In [52]:
# Combine results
results_all = pd.concat([results_system, results_tx], ignore_index = True)

#-------------POST PROCESS SIMULATION RESULTS---------------
# Post process interactions among strategies that affect the same variables
results_all_pp = cb.cb_process_interactions(results_all)

# SHIFT any stray costs incurred from 2015 to 2025 to 2025 and 2035
results_all_pp_shifted = cb.cb_shift_costs(results_all_pp)

Resolving Interactions in SCOE : TX:SCOE:INC_EFFICIENCY_APPLIANCE, TX:SCOE:SHIFT_FUEL_HEAT, TX:SCOE:SHIFT_FUEL_HEAT, TX:SCOE:DEC_DEMAND_HEAT 
Resolving Interactions in INEN : TX:INEN:SHIFT_FUEL_HEAT, TX:INEN:SHIFT_FUEL_HEAT, TX:INEN:SHIFT_FUEL_HEAT, TX:INEN:SHIFT_FUEL_HEAT, TX:INEN:SHIFT_FUEL_HEAT, TX:INEN:SHIFT_FUEL_HEAT, TX:INEN:INC_EFFICIENCY_ENERGY 
Resolving Interactions in SCOE : TX:SCOE:SHIFT_FUEL_HEAT, TX:SCOE:SHIFT_FUEL_HEAT, TX:SCOE:DEC_DEMAND_HEAT 
Resolving Interactions in INEN : TX:INEN:SHIFT_FUEL_HEAT, TX:INEN:SHIFT_FUEL_HEAT, TX:INEN:SHIFT_FUEL_HEAT, TX:INEN:SHIFT_FUEL_HEAT, TX:INEN:SHIFT_FUEL_HEAT, TX:INEN:SHIFT_FUEL_HEAT, TX:INEN:INC_EFFICIENCY_ENERGY 
Resolving Interactions in SCOE : TX:SCOE:SHIFT_FUEL_HEAT, TX:SCOE:SHIFT_FUEL_HEAT, TX:SCOE:DEC_DEMAND_HEAT 
Resolving Interactions in INEN : TX:INEN:SHIFT_FUEL_HEAT, TX:INEN:SHIFT_FUEL_HEAT, TX:INEN:SHIFT_FUEL_HEAT, TX:INEN:SHIFT_FUEL_HEAT, TX:INEN:SHIFT_FUEL_HEAT, TX:INEN:SHIFT_FUEL_HEAT, TX:INEN:INC_EFFICIENCY_ENERGY 


/Users/fabianfuentes/anaconda3/envs/ssp_uganda_env/lib/python3.11/site-packages/costs_benefits_ssp/cb_calculate.py:660: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  res_pre2025["variable"] = res_pre2025["variable"] + "_shifted" + (res_pre2025["time_period"]+SSP_GLOBAL_TIME_PERIOD_0).astype(str)#create a new variable so they can be recognized as shifted costs
/Users/fabianfuentes/anaconda3/envs/ssp_uganda_env/lib/python3.11/site-packages/costs_benefits_ssp/cb_calculate.py:661: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-

In [53]:
results_all_pp_shifted.head()

,strategy_code,future_id,region,time_period,difference_variable,variable_value_baseline,variable_value_pathway,difference_value,variable,value
0,PFLO:NZ,0,uganda,4,pop_unimproved_rural,1.932346e+07,1.932346e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.0
1,PFLO:NZ,0,uganda,5,pop_unimproved_rural,1.951974e+07,1.951974e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.0
2,PFLO:NZ,0,uganda,6,pop_unimproved_rural,1.896047e+07,1.896047e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.0
3,PFLO:NZ,0,uganda,7,pop_unimproved_rural,1.837001e+07,1.837001e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.0
4,PFLO:NZ,0,uganda,8,pop_unimproved_rural,1.774752e+07,1.774752e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.0


In [54]:
ssp_data.head()

,primary_id,region,time_period,area_agrc_crops_bevs_and_spices,area_agrc_crops_cereals,area_agrc_crops_fibers,area_agrc_crops_fruits,area_agrc_crops_herbs_and_other_perennial_crops,area_agrc_crops_nuts,area_agrc_crops_other_annual,...,yf_agrc_herbs_and_other_perennial_crops_tonne_ha,yf_agrc_nuts_tonne_ha,yf_agrc_other_annual_tonne_ha,yf_agrc_other_woody_perennial_tonne_ha,yf_agrc_pulses_tonne_ha,yf_agrc_rice_tonne_ha,yf_agrc_sugar_cane_tonne_ha,yf_agrc_tubers_tonne_ha,yf_agrc_vegetables_and_vines_tonne_ha,yf_lndu_supremum_pastures_tonne_per_ha
0,0,uganda,4,634960.593061,2.025945e+06,156073.139455,1.131815e+06,505.429083,513832.892470,2.915383e+06,...,2.172065,0.458120,1.528325,0.3674,0.799814,2.970174,77.726965,4.805888,4.035314,92.81
1,0,uganda,5,640126.452982,2.042427e+06,157342.906406,1.141024e+06,509.541112,518013.291655,2.939102e+06,...,2.196794,0.611231,1.681739,0.3674,0.805568,3.032345,78.023706,4.440105,4.013004,92.81
2,0,uganda,6,642767.262368,2.050853e+06,157992.016628,1.145731e+06,511.643198,520150.329354,2.951227e+06,...,2.210866,0.841393,2.293907,0.3674,0.807398,3.080626,61.430414,4.911305,4.023409,92.81
3,0,uganda,7,645322.479649,2.059006e+06,158620.088334,1.150285e+06,513.677153,522218.102852,2.962959e+06,...,2.225270,0.400800,2.699337,0.3674,0.810044,3.094956,77.255948,4.924276,4.030111,92.81
4,0,uganda,8,647798.760608,2.066907e+06,159228.757514,1.154699e+06,515.648274,524221.998246,2.974329e+06,...,2.225270,0.400800,2.699337,0.3674,0.810044,3.094956,77.255948,4.924276,4.030111,92.81


# Indoor air pollution benef

Health co-benefit from reduced household air pollution

To monetize the health benefits of reducing indoor air pollution from the use of traditional biomass stoves, we drew on estimates from the BAR-HAP model (The Benefits of Action to Reduce Household Air Pollution; WHO, 2021).
The study reports cost-effectiveness ratios ranging from 674 – 869 USD per DALY avoided for clean cooking interventions in Nepal, meaning that each avoided disability-adjusted life year (DALY) costs approximately 700 USD.

To adapt this metric to Uganda, we scaled the cost-effectiveness ratio by the relative difference in GDP per capita between Nepal (~1,450 USD) and Uganda (~1,070 USD), yielding an adjusted range of ≈ 500–640 USD per DALY avoided.
Assuming that the transition from solid fuels to clean cooking solutions avoids about 0.1 DALY per person per year, the implied annual health benefit equals 50–64 USD per person.

Given an average household size of five, this corresponds to an estimated 250–320 USD per household per year in health-related economic benefits.

This adjusted value was used in the cost–benefit analysis as the annual household health co-benefit of eliminating exposure to indoor air pollution from firewood and other solid fuels in Uganda.

In [55]:
ssp_data = ssp_data.merge(att_primary[['primary_id', 'strategy_id', 'future_id']], on='primary_id', how='left')
ssp_data = ssp_data.merge(att_strategy[['strategy_id', 'strategy_code']], on='strategy_id', how='left')

In [56]:
var_map = {
    "frac_scoe_heat_energy_residential_solid_biomass": "est_out_of_biomass",
    "frac_scoe_heat_energy_residential_electricity":   "est_out_of_elec",
    "frac_scoe_heat_energy_residential_hydrocarbon_gas_liquids":           "est_out_of_gas",
}

In [57]:
var_map_into = {
    "frac_scoe_heat_energy_residential_solid_biomass": "est_into_biomass",
    "frac_scoe_heat_energy_residential_electricity":  "est_into_elec",
    "frac_scoe_heat_energy_residential_hydrocarbon_gas_liquids":          "est_into_gas",
}

In [58]:
def add_into(
    df: pd.DataFrame,
    time_col: str,
    group_cols: list[str],
    var_map: dict[str, str],
) -> pd.DataFrame:
    """
    into(t) = max(frac(t+1) - frac(t), 0)
    """
    df = df.sort_values(group_cols + [time_col]).copy()

    for frac_var, into_var in var_map.items():
        next_frac = df.groupby(group_cols, sort=False)[frac_var].shift(-1)
        df[into_var] = (next_frac - df[frac_var]).clip(lower=0).fillna(0)

    return df

In [59]:
def add_out_of(
    df: pd.DataFrame,
    time_col: str,
    group_cols: list[str],
    var_map: dict[str, str],
) -> pd.DataFrame:
    """
    out_of(t) = max(frac(t) - frac(t+1), 0)
    """
    df = df.sort_values(group_cols + [time_col]).copy()

    for frac_var, out_var in var_map.items():
        next_frac = df.groupby(group_cols, sort=False)[frac_var].shift(-1)
        df[out_var] = (df[frac_var] - next_frac).clip(lower=0).fillna(0)

    return df

In [60]:
df_ssp = add_out_of(
    df=ssp_data,
    time_col="time_period",                
    group_cols=["strategy_id"],   
    var_map=var_map,
)

In [61]:
df_ssp = add_into(
    df=df_ssp,
    time_col="time_period",            
    group_cols=["strategy_id", "region"],
    var_map=var_map_into,
)

In [62]:
df_ssp[['time_period','strategy_id','frac_scoe_heat_energy_residential_solid_biomass','est_out_of_biomass', 'est_into_biomass']].head(7)

,time_period,strategy_id,frac_scoe_heat_energy_residential_solid_biomass,est_out_of_biomass,est_into_biomass
0,4,0,0.960324,0.027545,0.0
1,5,0,0.932779,0.031764,0.0
2,6,0,0.901015,0.000000,0.0
3,7,0,0.901015,0.000000,0.0
4,8,0,0.901015,0.000000,0.0
5,9,0,0.901015,0.000000,0.0
6,10,0,0.901015,0.000000,0.0


In [63]:
df_ssp[['time_period','strategy_id','frac_scoe_heat_energy_residential_electricity','est_out_of_elec', 'est_into_elec']].head(7)

,time_period,strategy_id,frac_scoe_heat_energy_residential_electricity,est_out_of_elec,est_into_elec
0,4,0,0.023887,0.0,0.014894
1,5,0,0.038781,0.0,0.017394
2,6,0,0.056174,0.0,0.000000
3,7,0,0.056174,0.0,0.000000
4,8,0,0.056174,0.0,0.000000
5,9,0,0.056174,0.0,0.000000
6,10,0,0.056174,0.0,0.000000


In [64]:
df_ssp[['time_period','strategy_id','frac_scoe_heat_energy_residential_hydrocarbon_gas_liquids','est_out_of_gas', 'est_into_gas']].head(7)

,time_period,strategy_id,frac_scoe_heat_energy_residential_hydrocarbon_gas_liquids,est_out_of_gas,est_into_gas
0,4,0,0.011962,0.0,0.009584
1,5,0,0.021546,0.0,0.010887
2,6,0,0.032432,0.0,0.000000
3,7,0,0.032432,0.0,0.000000
4,8,0,0.032432,0.0,0.000000
5,9,0,0.032432,0.0,0.000000
6,10,0,0.032432,0.0,0.000000


In [65]:
def add_transitions(df: pd.DataFrame) -> pd.DataFrame:
    df = df.copy()

    # 1. Biomass -> Gas
    df["biomass_to_gas"] = df[["est_out_of_biomass", "est_into_gas"]].min(axis=1)

    # Remaining out of biomass after going to gas
    rem_out_biomass = df["est_out_of_biomass"] - df["biomass_to_gas"]

    # Remaining into gas after receiving biomass
    rem_into_gas = df["est_into_gas"] - df["biomass_to_gas"]

    # 2. Biomass -> Elec
    df["biomass_to_elec"] = pd.concat(
        [rem_out_biomass, df["est_into_elec"]],
        axis=1
    ).min(axis=1)

    # Remaining into elec after biomass
    rem_into_elec = df["est_into_elec"] - df["biomass_to_elec"]

    # 3. Gas -> Elec
    df["gas_to_elec"] = pd.concat(
        [df["est_out_of_gas"], rem_into_elec],
        axis=1
    ).min(axis=1)

    return df

In [66]:
df_ssp = add_transitions(df_ssp)

In [67]:
df_ssp.strategy_id.value_counts()

strategy_id
0       52
6004    52
6008    52
6009    52
Name: count, dtype: int64

In [68]:
df_ssp[df_ssp.strategy_id == 0][['time_period','strategy_id','biomass_to_gas','biomass_to_elec','gas_to_elec']].head(15)

,time_period,strategy_id,biomass_to_gas,biomass_to_elec,gas_to_elec
0,4,0,0.009584,0.014894,0.0
1,5,0,0.010887,0.017394,0.0
2,6,0,0.000000,0.000000,0.0
3,7,0,0.000000,0.000000,0.0
4,8,0,0.000000,0.000000,0.0
5,9,0,0.000000,0.000000,0.0
6,10,0,0.000000,0.000000,0.0
7,11,0,0.008477,0.000000,0.0
8,12,0,0.012615,0.000000,0.0
9,13,0,0.018555,0.000000,0.0


In [69]:
df_ssp[df_ssp.strategy_id == 6004][['time_period','strategy_id','biomass_to_gas','biomass_to_elec','gas_to_elec']].head(15)

,time_period,strategy_id,biomass_to_gas,biomass_to_elec,gas_to_elec
52,4,6004,0.009584,0.014894,0.0
53,5,6004,0.010887,0.017394,0.0
54,6,6004,0.000000,0.000000,0.0
55,7,6004,0.000000,0.000000,0.0
56,8,6004,0.000000,0.000000,0.0
57,9,6004,0.000000,0.000000,0.0
58,10,6004,0.000000,0.000000,0.0
59,11,6004,0.008477,0.000000,0.0
60,12,6004,0.011370,0.021708,0.0
61,13,6004,0.016447,0.021708,0.0


In [70]:
df_ssp[df_ssp.strategy_id == 6008][['time_period','strategy_id','biomass_to_gas','biomass_to_elec','gas_to_elec']].head(15)

,time_period,strategy_id,biomass_to_gas,biomass_to_elec,gas_to_elec
104,4,6008,0.009584,0.014894,0.0
105,5,6008,0.010887,0.017394,0.0
106,6,6008,0.000000,0.000000,0.0
107,7,6008,0.000000,0.000000,0.0
108,8,6008,0.000000,0.000000,0.0
109,9,6008,0.000000,0.000000,0.0
110,10,6008,0.000000,0.000000,0.0
111,11,6008,0.008477,0.000000,0.0
112,12,6008,0.012615,0.000000,0.0
113,13,6008,0.018555,0.000000,0.0


In [71]:
df_ssp[df_ssp.strategy_id == 6009][['time_period','strategy_id','biomass_to_gas','biomass_to_elec','gas_to_elec']].head(15)

,time_period,strategy_id,biomass_to_gas,biomass_to_elec,gas_to_elec
156,4,6009,0.009584,0.014894,0.0
157,5,6009,0.010887,0.017394,0.0
158,6,6009,0.000000,0.000000,0.0
159,7,6009,0.000000,0.000000,0.0
160,8,6009,0.000000,0.000000,0.0
161,9,6009,0.000000,0.000000,0.0
162,10,6009,0.000000,0.000000,0.0
163,11,6009,0.008477,0.000000,0.0
164,12,6009,0.012615,0.000000,0.0
165,13,6009,0.018555,0.000000,0.0


In [72]:
cost_benfit = df_ssp[['time_period','strategy_id','population_gnrl_total','biomass_to_gas','biomass_to_elec']]

In [73]:
df_factors = pd.DataFrame(
    [
        # ======================
        # biomass_to_gas
        # ======================
        {
            "transition": "biomass_to_gas",
            "policy": "subsidy_for_stove",
            "cost_per_capita_ppp_2019": -1.349440118,
            "benefit_per_capita_ppp_2019": 4.767680737,
        },
        {
            "transition": "biomass_to_gas",
            "policy": "fuel_subsidy",
            "cost_per_capita_ppp_2019": -3.431367796,
            "benefit_per_capita_ppp_2019": 5.341173225,
        },
        {
            "transition": "biomass_to_gas",
            "policy": "financing",
            "cost_per_capita_ppp_2019": -3.368400432,
            "benefit_per_capita_ppp_2019": 12.44425028,
        },
        {
            "transition": "biomass_to_gas",
            "policy": "intensive_behavioral_change_campaign",
            "cost_per_capita_ppp_2019": -2.012617193,
            "benefit_per_capita_ppp_2019": 6.868823172,
        },

        # ======================
        # biomass_to_electricity
        # ======================
        {
            "transition": "biomass_to_elec",
            "policy": "subsidy_for_stove",
            "cost_per_capita_ppp_2019": -2.659697106,
            "benefit_per_capita_ppp_2019": 9.566241834,
        },
        {
            "transition": "biomass_to_elec",
            "policy": "fuel_subsidy",
            "cost_per_capita_ppp_2019": -6.229455845,
            "benefit_per_capita_ppp_2019": 11.33011101,
        },
        {
            "transition": "biomass_to_elec",
            "policy": "financing",
            "cost_per_capita_ppp_2019": -4.838372823,
            "benefit_per_capita_ppp_2019": 17.56115615,
        },
        {
            "transition": "biomass_to_elec",
            "policy": "intensive_behavioral_change_campaign",
            "cost_per_capita_ppp_2019": -3.535655711,
            "benefit_per_capita_ppp_2019": 11.5649704,
        }
    ]
)

In [74]:
transitions = [
    "biomass_to_gas",
    "biomass_to_elec"
]

ssp_long = (
    df_ssp
    .melt(
        id_vars=["time_period",'strategy_id',"population_gnrl_total"],
        value_vars=transitions,
        var_name="transition",
        value_name="transition_frac"
    )
)

In [75]:
ssp_long["population_affected"] = (
    ssp_long["transition_frac"] * ssp_long["population_gnrl_total"]
)

In [76]:
cb = ssp_long.merge(df_factors, on="transition", how="left")

cb["total_cost"] = (
    cb["population_affected"] * cb["cost_per_capita_ppp_2019"]
)

cb["total_benefit"] = (
    cb["population_affected"] * cb["benefit_per_capita_ppp_2019"]
)

cb["net_benefit"] = cb["total_benefit"] - cb["total_cost"]

In [77]:
cb

,time_period,strategy_id,population_gnrl_total,transition,transition_frac,population_affected,policy,cost_per_capita_ppp_2019,benefit_per_capita_ppp_2019,total_cost,total_benefit,net_benefit
0,4,0,39853824,biomass_to_gas,0.009584,381955.329412,subsidy_for_stove,-1.349440,4.767681,-5.154258e+05,1.821041e+06,2.336467e+06
1,4,0,39853824,biomass_to_gas,0.009584,381955.329412,fuel_subsidy,-3.431368,5.341173,-1.310629e+06,2.040090e+06,3.350719e+06
2,4,0,39853824,biomass_to_gas,0.009584,381955.329412,financing,-3.368400,12.444250,-1.286578e+06,4.753148e+06,6.039726e+06
3,4,0,39853824,biomass_to_gas,0.009584,381955.329412,intensive_behavioral_change_campaign,-2.012617,6.868823,-7.687299e+05,2.623584e+06,3.392313e+06
4,5,0,40996690,biomass_to_gas,0.010887,446317.967983,subsidy_for_stove,-1.349440,4.767681,-6.022794e+05,2.127902e+06,2.730181e+06
...,...,...,...,...,...,...,...,...,...,...,...,...
1659,54,6009,116777146,biomass_to_elec,0.000000,0.000000,intensive_behavioral_change_campaign,-3.535656,11.564970,-0.000000e+00,0.000000e+00,0.000000e+00
1660,55,6009,118356823,biomass_to_elec,0.000000,0.000000,subsidy_for_stove,-2.659697,9.566242,-0.000000e+00,0.000000e+00,0.000000e+00
1661,55,6009,118356823,biomass_to_elec,0.000000,0.000000,fuel_subsidy,-6.229456,11.330111,-0.000000e+00,0.000000e+00,0.000000e+00
1662,55,6009,118356823,biomass_to_elec,0.000000,0.000000,financing,-4.838373,17.561156,-0.000000e+00,0.000000e+00,0.000000e+00


In [78]:
cb_long = (
    cb
    .melt(
        id_vars=[
            "time_period",
            "strategy_id",
            "transition",
            "policy",
            "population_affected",
        ],
        value_vars=["total_cost", "total_benefit"],
        var_name="metric_type",
        value_name="value",
    )
)

In [79]:
cb_long["metric_type"] = pd.Categorical(
    cb_long["metric_type"],
    categories=["total_benefit", "total_cost"], 
    ordered=True,
)

cb_long = cb_long.sort_values(
    by=[
        "strategy_id",
        "transition",
        "policy",
        "metric_type",   # 👈 primero benefit, luego cost
        "time_period",   # 👈 dentro de cada bloque, tiempo ascendente
    ],
    ascending=[
        True,
        True,
        True,
        True,
        True,
    ],
).reset_index(drop=True)

In [80]:
cb_long[cb_long.strategy_id==0].head(10)

,time_period,strategy_id,transition,policy,population_affected,metric_type,value
0,4,0,biomass_to_elec,financing,593584.192686,total_benefit,1.042402e+07
1,5,0,biomass_to_elec,financing,713087.735044,total_benefit,1.252265e+07
2,6,0,biomass_to_elec,financing,0.000000,total_benefit,0.000000e+00
3,7,0,biomass_to_elec,financing,0.000000,total_benefit,0.000000e+00
4,8,0,biomass_to_elec,financing,0.000000,total_benefit,0.000000e+00
5,9,0,biomass_to_elec,financing,0.000000,total_benefit,0.000000e+00
6,10,0,biomass_to_elec,financing,0.000000,total_benefit,0.000000e+00
7,11,0,biomass_to_elec,financing,0.000000,total_benefit,0.000000e+00
8,12,0,biomass_to_elec,financing,0.000000,total_benefit,0.000000e+00
9,13,0,biomass_to_elec,financing,0.000000,total_benefit,0.000000e+00


In [81]:
cb_long[cb_long.strategy_id==6004].head(10)

,time_period,strategy_id,transition,policy,population_affected,metric_type,value
832,4,6004,biomass_to_elec,financing,5.935842e+05,total_benefit,1.042402e+07
833,5,6004,biomass_to_elec,financing,7.130877e+05,total_benefit,1.252265e+07
834,6,6004,biomass_to_elec,financing,0.000000e+00,total_benefit,0.000000e+00
835,7,6004,biomass_to_elec,financing,0.000000e+00,total_benefit,0.000000e+00
836,8,6004,biomass_to_elec,financing,0.000000e+00,total_benefit,0.000000e+00
837,9,6004,biomass_to_elec,financing,0.000000e+00,total_benefit,0.000000e+00
838,10,6004,biomass_to_elec,financing,0.000000e+00,total_benefit,0.000000e+00
839,11,6004,biomass_to_elec,financing,0.000000e+00,total_benefit,0.000000e+00
840,12,6004,biomass_to_elec,financing,1.083846e+06,total_benefit,1.903359e+07
841,13,6004,biomass_to_elec,financing,1.114018e+06,total_benefit,1.956345e+07


In [82]:
cb_long[cb_long.strategy_id==6008].head()

,time_period,strategy_id,transition,policy,population_affected,metric_type,value
1664,4,6008,biomass_to_elec,financing,593584.192686,total_benefit,1.042402e+07
1665,5,6008,biomass_to_elec,financing,713087.735044,total_benefit,1.252265e+07
1666,6,6008,biomass_to_elec,financing,0.000000,total_benefit,0.000000e+00
1667,7,6008,biomass_to_elec,financing,0.000000,total_benefit,0.000000e+00
1668,8,6008,biomass_to_elec,financing,0.000000,total_benefit,0.000000e+00


In [83]:
cb_long[cb_long.strategy_id==6009].head()

,time_period,strategy_id,transition,policy,population_affected,metric_type,value
2496,4,6009,biomass_to_elec,financing,593584.192686,total_benefit,1.042402e+07
2497,5,6009,biomass_to_elec,financing,713087.735044,total_benefit,1.252265e+07
2498,6,6009,biomass_to_elec,financing,0.000000,total_benefit,0.000000e+00
2499,7,6009,biomass_to_elec,financing,0.000000,total_benefit,0.000000e+00
2500,8,6009,biomass_to_elec,financing,0.000000,total_benefit,0.000000e+00


In [84]:
cb_long['value'] = cb_long['value'] /1e6

In [85]:
# Save the results
cb_long.to_csv(os.path.join(CB_OUTPUT, "cba_air_pollution.csv"), index = False)

In [86]:
CB_OUTPUT

'/Users/fabianfuentes/git/ssp_uganda_data/ssp_modeling/cb/cb_results'